activation/activation.go

```diff
@@ -159,80 +170,136 @@ func (b *Builder) loop() {
+	if b.prevATX == nil {
+		prevAtxId, err := b.GetPrevAtxId(b.nodeId)
+		if err != nil {
+			b.log.Info("no prev ATX found, starting fresh")
+		} else {
+			b.prevATX, err = b.db.GetAtx(*prevAtxId)
+			if err != nil {
+				// TODO: handle inconsistent state
+				b.log.Panic("prevAtx (id: %v) not found in DB -- inconsistent state", prevAtxId.ShortId())
+			}
+		}
+	}
+	seq := uint64(0)
+	prevAtxId := *types.EmptyAtxId
+	if b.prevATX != nil {
+		seq = b.prevATX.Sequence + 1
+		//todo: handle this case for loading mem and recovering
+		//check if this node hasn't published an activation already
+		if b.prevATX.PubLayerIdx.GetEpoch(b.layersPerEpoch) == epoch+1 {
+			b.log.Info("atx already created for epoch %v, aborting", epoch)
+			return fmt.Errorf("atx already created for epoch %v, aborting", epoch)
+		}
+		prevAtxId = b.prevATX.Id()
+	}
+	posAtxEndTick := uint64(0)
+	b.posLayerID = types.LayerID(0)
+
+	//positioning atx is from this epoch as well, since we will be publishing the atx in the next epoch
+	//todo: what if no other atx was received in this epoch yet?
+	posAtxId := *types.EmptyAtxId
+	posAtx, err := b.GetPositioningAtx(epoch)
+	atxPubLayerID := types.LayerID(0)
+	if err == nil {
+		posAtxEndTick = posAtx.EndTick
+		b.posLayerID = posAtx.PubLayerIdx
+		atxPubLayerID = b.posLayerID.Add(b.layersPerEpoch)
+		posAtxId = posAtx.Id()
+	} else {
+		if !epoch.IsGenesis() {
+			return fmt.Errorf("cannot find pos atx: %v", err)
+		}
+	}
+
+	b.challenge = &types.NIPSTChallenge{
+		NodeId:         b.nodeId,
+		Sequence:       seq,
+		PrevATXId:      prevAtxId,
+		PubLayerIdx:    atxPubLayerID,
+		StartTick:      posAtxEndTick,
+		EndTick:        b.tickProvider.NumOfTicks(), //todo: add provider when
+		PositioningAtx: posAtxId,
+	}
+	err = b.storeChallenge(b.challenge)
+	if err != nil {
+		log.Error("challenge cannot be stored: %s", err)
+	}
+	return nil
+}
+
+func (b Builder) getNipstKey() []byte {
+	return []byte("nipst")
+}
+
+func (b *Builder) storeChallenge(ch *types.NIPSTChallenge) error {
+	bts, err := types.InterfaceToBytes(ch)
+	if err != nil {
+		return err
+	}
+	return b.store.Put(b.getNipstKey(), bts)
+}
+
+func (b *Builder) loadChallenge() error {
+	bts, err := b.store.Get(b.getNipstKey())
+	if err != nil {
+		return err
+	}
+	if len(bts) > 0 {
+		tp := &types.NIPSTChallenge{}
+		err = types.BytesToInterface(bts, tp)
+		if err != nil {
+			return err
+		}
+		b.challenge = tp
+	}
+	return nil
+}
+
+func (b *Builder) discardChallenge() error {
+	return b.store.Put(b.getNipstKey(), []byte{})
+}
+
 // PublishActivationTx attempts to publish an atx for the given epoch, it returns an error if an atx cannot be created
 // publish atx may not produce an atx each time it is called, that is expected behaviour as well.
-func (b *Builder) PublishActivationTx(epoch types.EpochId) (bool, error) {
+func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
 	if b.nipst != nil {
 		b.log.Info("re-entering atx creation in epoch %v", epoch)
```

Comment:
If it's not too much of a hassle, please change this log (and the one below) to something like "... in epoch j, for epoch k"

Reply:
There is a print that says this exactly if we have a challenge